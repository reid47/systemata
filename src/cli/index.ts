#!/usr/bin/env node
import * as path from 'path';
import * as fs from 'fs';
import { build } from './build';
import { version } from '../../package.json';
import { parseArgs, knownCommands } from './parse-args';
import { usage } from './usage';

const error = (message: string) => {
  console.error(`${message}\n\nUse '-h' or '--help' to show usage information.`);
  process.exit(1);
};

const { args, command } = parseArgs(process.argv.slice(2));

if (args.help) {
  console.log(usage);
  process.exit(0);
}

if (args.version) {
  console.log(version);
  process.exit(0);
}

if (command === 'init') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  if (fs.existsSync(path.resolve(process.cwd(), configFile as string))) {
    error(`File '${configFile}' already exists. Please specify a file name that doesn't exist.`);
  }
}

if (command === 'build') {
  const configFile = args['-c'] || args['--config'];

  if (!configFile || typeof configFile !== 'string') {
    error("No config file specified. Use '-c' or '--config' to pass a path to a system config file.");
  }

  let config;
  try {
    config = require(path.resolve(process.cwd(), configFile as string));
  } catch (err) {
    error(`Error loading config file: ${err.message}`);
  }

  try {
    build(config);
    process.exit(0);
  } catch (err) {
    error(err);
  }
}

const validCommands = Object.keys(knownCommands).join(', ');

if (!command) {
  error(`No command specified. Command should be one of: ${validCommands}`);
}

error(`Invalid command specified: ${command}\nCommand should be one of: ${validCommands}`);
