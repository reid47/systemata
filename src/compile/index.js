import fs from 'fs';
import path from 'path';
import * as scss from './to-scss';

export default function compile(config) {
  const ext = scss.extension;
  fs.writeFileSync(
    path.resolve(__dirname, '../../ignored/wow' + ext),
    scss.compile(config),
    'utf8'
  );
}
