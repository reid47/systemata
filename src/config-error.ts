export class ConfigurationError extends Error {
  constructor(configPath: string, message: string) {
    const fullMessage = `Invalid configuration at "${configPath}": ${message}`;
    super(fullMessage);
  }
}
