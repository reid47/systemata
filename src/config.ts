import {
  NamespaceConfig,
  Config,
  ColorConfig,
  PropertyMappingConfig,
  SpacingConfig,
  OutputConfig,
  SystemConfig
} from './types';

function resolveSystem(systemConfig: any): SystemConfig {
  systemConfig = systemConfig || {};

  return {
    name: systemConfig.name || 'untitled-system',
    version: systemConfig.version || 'version',
    description: systemConfig.description || ''
  };
}

function resolveOutput(outputConfig: any): OutputConfig {
  return {
    format: outputConfig.format || 'css'
  };
}

function resolveNamespace(namespaceConfig: any): NamespaceConfig {
  namespaceConfig = namespaceConfig || {};

  return {
    prefix: namespaceConfig.prefix || false,
    parentClass: namespaceConfig.parentClass || false,
    peerClass: namespaceConfig.peerClass || false
  };
}

function resolvePropertyMapping(mappingsConfig: any): PropertyMappingConfig {
  mappingsConfig = (mappingsConfig || {}) as object;

  const defaultMappings: PropertyMappingConfig = {
    'background-color': 'bg',
    'border-color': 'brc',
    color: 'c',
    fill: 'fill',
    margin: 'm',
    'margin-bottom': 'mb',
    'margin-left': 'ml',
    'margin-right': 'mr',
    'margin-top': 'mt',
    'margin-top,margin-bottom': 'mv',
    'margin-left,margin-right': 'mh',
    padding: 'p',
    'padding-bottom': 'pb',
    'padding-left': 'pl',
    'padding-right': 'pr',
    'padding-top': 'pt',
    'padding-top,padding-bottom': 'pv',
    'padding-left,padding-right': 'ph'
  };

  return { ...defaultMappings, ...mappingsConfig };
}

function resolveColors(colorConfig: any): ColorConfig {
  if (!colorConfig) return {};
  return colorConfig as ColorConfig;
}

function resolveSpacing(spacingConfig: any): SpacingConfig {
  if (!spacingConfig) return {};
  return spacingConfig as SpacingConfig;
}

export function resolveConfig(config: any): Config {
  return {
    system: resolveSystem(config.system),
    namespace: resolveNamespace(config.namespace),
    propertyMapping: resolvePropertyMapping(config.propertyMapping),
    output: resolveOutput(config.output),
    color: resolveColors(config.color),
    spacing: resolveSpacing(config.spacing)
  };
}
