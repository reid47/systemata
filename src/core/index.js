const { CssBuilder } = require('../css-builder');

function generate(config) {
  const { typography } = config;

  const cssBuilder = new CssBuilder();

  Object.keys(typography.sizes).forEach(typeSize => {
    cssBuilder.addRule(typeSize, {
      'font-size': typography.sizes[typeSize]
    });
  });

  console.log(JSON.stringify(config, null, 2));
  console.log(cssBuilder.toCss());
}

module.exports = { generate };
