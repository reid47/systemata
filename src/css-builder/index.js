class CssBuilder {
  constructor() {
    this.rules = {};
  }

  addRule(selector, styles) {
    this.rules[selector] = {
      ...this.rules[selector],
      ...styles
    };
  }

  toCss() {
    const lines = [];

    Object.keys(this.rules).forEach(selector => {
      lines.push(`${selector} {`);
      Object.keys(this.rules[selector]).forEach(property => {
        const value = this.rules[selector][property];
        lines.push(`  ${property}: ${value};`);
      });
      lines.push('}\n');
    });

    return lines.join('\n');
  }
}

module.exports = { CssBuilder };
