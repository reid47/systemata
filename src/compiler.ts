import { resolveConfig } from './config';
import { Config, OutputOptions } from './types';
import { generateCss } from './generate-css';

export class Compiler {
  config: Config;
  output: OutputOptions;

  constructor(config: any, output: OutputOptions) {
    this.config = resolveConfig(config);
    this.output = output;
  }

  toString() {
    if (this.output.format === 'css') return generateCss(this.config);

    throw new Error('Unsupported output configuration');
  }
}
