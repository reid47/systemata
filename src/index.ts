import { Compiler } from './compiler';
import { OutputOptions } from './types';

export function generate(config: any, output: any): string {
  return new Compiler(config, output).toString();
}
