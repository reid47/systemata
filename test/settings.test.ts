import { generate } from '../src/index';

const output = { format: 'css' };

describe('namespace', () => {
  test('namespace (object with prefix)', () => {
    const config = {
      output,
      namespace: {
        prefix: 'wow-'
      },
      color: {
        primary: '#f00',
        secondary: '#0f0'
      }
    };

    expect(generate(config)).toMatchSnapshot();
  });

  test('namespace (object with parentClass)', () => {
    const config = {
      output,
      namespace: {
        parentClass: 'parent'
      },
      color: {
        primary: '#f00',
        secondary: '#0f0'
      }
    };

    expect(generate(config)).toMatchSnapshot();
  });

  test('namespace (object with peerClass)', () => {
    const config = {
      output,
      namespace: { peerClass: 'peer' },
      color: {
        primary: '#f00',
        secondary: '#0f0'
      }
    };

    expect(generate(config)).toMatchSnapshot();
  });

  test('namespace (object with multiple namespace types)', () => {
    const config = {
      output,
      namespace: {
        prefix: 'pfx-',
        parentClass: 'parent',
        peerClass: 'peer'
      },
      color: {
        primary: '#f00',
        secondary: '#0f0'
      }
    };

    expect(generate(config)).toMatchSnapshot();
  });
});

describe('propertyMapping', () => {
  test('renaming properties', () => {
    const config = {
      output,
      propertyMapping: {
        color: 'my-name-for-color'
      },
      color: {
        primary: '#f00',
        secondary: '#0f0'
      }
    };

    expect(generate(config)).toMatchSnapshot();
  });

  test('excluding properties', () => {
    const config = {
      output,
      propertyMapping: {
        fill: false
      },
      color: {
        primary: '#f00',
        secondary: '#0f0'
      }
    };

    expect(generate(config)).toMatchSnapshot();
  });
});
