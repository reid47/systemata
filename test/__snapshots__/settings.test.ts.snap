// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`namespace namespace (object with multiple namespace types) 1`] = `
".parent .peer.pfx-bg-primary { background-color: #f00 }
.parent .peer.pfx-bg-secondary { background-color: #0f0 }
.parent .peer.pfx-brc-primary { border-color: #f00 }
.parent .peer.pfx-brc-secondary { border-color: #0f0 }
.parent .peer.pfx-c-primary { color: #f00 }
.parent .peer.pfx-c-secondary { color: #0f0 }
.parent .peer.pfx-fill-primary { fill: #f00 }
.parent .peer.pfx-fill-secondary { fill: #0f0 }"
`;

exports[`namespace namespace (object with parentClass) 1`] = `
".parent .bg-primary { background-color: #f00 }
.parent .bg-secondary { background-color: #0f0 }
.parent .brc-primary { border-color: #f00 }
.parent .brc-secondary { border-color: #0f0 }
.parent .c-primary { color: #f00 }
.parent .c-secondary { color: #0f0 }
.parent .fill-primary { fill: #f00 }
.parent .fill-secondary { fill: #0f0 }"
`;

exports[`namespace namespace (object with peerClass) 1`] = `
".peer.bg-primary { background-color: #f00 }
.peer.bg-secondary { background-color: #0f0 }
.peer.brc-primary { border-color: #f00 }
.peer.brc-secondary { border-color: #0f0 }
.peer.c-primary { color: #f00 }
.peer.c-secondary { color: #0f0 }
.peer.fill-primary { fill: #f00 }
.peer.fill-secondary { fill: #0f0 }"
`;

exports[`namespace namespace (object with prefix) 1`] = `
".wow-bg-primary { background-color: #f00 }
.wow-bg-secondary { background-color: #0f0 }
.wow-brc-primary { border-color: #f00 }
.wow-brc-secondary { border-color: #0f0 }
.wow-c-primary { color: #f00 }
.wow-c-secondary { color: #0f0 }
.wow-fill-primary { fill: #f00 }
.wow-fill-secondary { fill: #0f0 }"
`;

exports[`propertyMapping excluding properties 1`] = `
".bg-primary { background-color: #f00 }
.bg-secondary { background-color: #0f0 }
.brc-primary { border-color: #f00 }
.brc-secondary { border-color: #0f0 }
.c-primary { color: #f00 }
.c-secondary { color: #0f0 }"
`;

exports[`propertyMapping renaming properties 1`] = `
".bg-primary { background-color: #f00 }
.bg-secondary { background-color: #0f0 }
.brc-primary { border-color: #f00 }
.brc-secondary { border-color: #0f0 }
.my-name-for-color-primary { color: #f00 }
.my-name-for-color-secondary { color: #0f0 }
.fill-primary { fill: #f00 }
.fill-secondary { fill: #0f0 }"
`;
